### ISSUE TITLE
No tracing context when processing ClientResponse body

### ISSUE THREAD

**Issue by norbson** [spring-projects/spring-framework#31817]
When processing ClientResponse body, tracing context is missing.

One of many way to reproduce it:
```java
clientResponse.mutate().body(dataBuffer -> dataBuffer.doOnNext(db -> log.info("No tracing here")))
```
Environment:
- Spring Boot 3.1.5
- Spring Framework: 6.0.13
- Project Reactor 3.5.11

**Comment by norbson** [comment]
I will prepare a sample. This is only a snippet. I have enabled automatic context propagation and there is active observation on webClient call. So i expect existence of traceId and spanId during body transformation.

**Comment by norbson** [comment]
Sample project
https://github.com/norbson/sb-tracing

As You see, during response body streaming, tracing is lost. 
I expect to be able to trace processing status.


![image](https://github.com/spring-projects/spring-framework/assets/2744731/153000d5-580b-4a8a-95c9-b4bc8dd3d1f4)

**Comment by norbson** [comment]
Even more obvious case with Flux. 

Code:
https://github.com/norbson/sb-tracing/tree/flux

In this case tracing context is missing also later in the chain, when processing already decoded element.

<img width="1308" alt="image" src="https://github.com/spring-projects/spring-framework/assets/2744731/be7129bc-c941-43d1-b36d-f88353b551df">

**Comment by norbson** [comment]
In summary, I believe that restoring threadLocal failed in these cases.

**Comment by bclozel** [comment]
I've been having a look and I'm not sure at this stage if this is a bug in Spring Framework, in Reactor or an expected behavior given the decoder setup.

I'm seeing that the context is not being established in the `tokens.handle` part of the `AbstractJackson2Decoder#decode`. I'll need some more time to pinpoint where the context is "lost" and whether we can do something about it.

**Comment by bclozel** [comment]
After chatting with the reactor team (thanks a lot @chemicL !), this is actually another instance of #30273. This is a complex issue that was fixed in Reactor 3.6 (see [the dedicated reactor announcement blog post](https://spring.io/blog/2023/10/31/what-new-is-coming-in-reactor-core-3-6-0#better-automatic-context-propagation)).

I've upgraded your sample to Spring Boot 3.2 and it all works fine. This is a major feature in Reactor and will not be backported to previous generations.

I'm closing this issue as a result. Thanks for the report!
